File: source/jugglinglab/prop/Prop.java
// Prop.java
//
// Copyright 2019 by Jack Boyce (jboyce@gmail.com)

package jugglinglab.prop;

import java.awt.*;
import java.text.MessageFormat;
import java.util.ResourceBundle;

import jugglinglab.util.*;


public abstract class Prop {
    static final ResourceBundle guistrings = jugglinglab.JugglingLab.guistrings;
    static final ResourceBundle errorstrings = jugglinglab.JugglingLab.errorstrings;

    protected String initString;

    public static final String[] builtinProps = { "Ball", "Image", "Ring" };

    // Creates a new prop of the given type
    public static Prop newProp(String type) throws JuggleExceptionUser {
        if (type == null)
            throw new JuggleExceptionUser("Prop type not specified");

        if (type.equalsIgnoreCase("ball"))
            return new BallProp();
        else if (type.equalsIgnoreCase("image"))
            return new ImageProp();
        else if (type.equalsIgnoreCase("ring"))
            return new RingProp();

        String template = errorstrings.getString("Error_prop_type");
        Object[] arguments = { type };
        throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
    }

    public void initProp(String st) throws JuggleExceptionUser {
        initString = st;
        this.init(st);
    }

    public abstract String getType();

    public abstract Color getEditorColor();

    public abstract ParameterDescriptor[] getParameterDescriptors();

    protected abstract void init(String st) throws JuggleExceptionUser;
    public abstract Coordinate getMax();    // in cm
    public abstract Coordinate getMin();    // in cm
    public abstract double getWidth();      // prop width in cm
    public abstract Image getProp2DImage(double zoom, double[] camangle);
    public abstract Dimension getProp2DSize(double zoom);
    public abstract Dimension getProp2DCenter(double zoom);
    public abstract Dimension getProp2DGrip(double zoom);

    /*
    public abstract Object getPropIDX3D();
    public abstract Coordinate getPropIDX3DGrip();
    */
}

File: composeApp/src/commonMain/kotlin/jugglinglab/prop/Prop.kt
//
// Prop.kt
//
// This is the base type of all props in Juggling Lab.
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.prop

import jugglinglab.composeapp.generated.resources.*
import jugglinglab.util.Coordinate
import jugglinglab.util.JuggleExceptionUser
import jugglinglab.util.ParameterDescriptor
import jugglinglab.util.getStringResource
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.IntSize
import androidx.compose.ui.graphics.ImageBitmap
import org.jetbrains.compose.resources.StringResource

abstract class Prop {
    protected var initString: String? = null

    @Throws(JuggleExceptionUser::class)
    fun initProp(st: String?) {
        initString = st
        init(st)
    }

    //--------------------------------------------------------------------------
    // Abstract methods defined by subclasses
    //--------------------------------------------------------------------------

    abstract val type: String

    abstract val isColorable: Boolean

    abstract fun getEditorColor(): Color

    abstract val parameterDescriptors: List<ParameterDescriptor>

    @Throws(JuggleExceptionUser::class)
    protected abstract fun init(st: String?)

    abstract fun getMax(): Coordinate? // in cm

    abstract fun getMin(): Coordinate? // in cm

    abstract fun getWidth(): Double // prop width in cm, in juggler space

    abstract fun getProp2DImage(zoom: Double, camangle: DoubleArray): ImageBitmap?

    abstract fun getProp2DSize(zoom: Double, camangle: DoubleArray): IntSize?

    abstract fun getProp2DCenter(zoom: Double, camangle: DoubleArray): IntSize?

    abstract fun getProp2DGrip(zoom: Double, camangle: DoubleArray): IntSize?

    companion object {
        val builtinProps: List<String> = listOf(
            "Ball",
            "Image",
            "Ring",
        )
        val builtinPropsStringResources: List<StringResource> = listOf(
            Res.string.gui_prop_name_ball,
            Res.string.gui_prop_name_image,
            Res.string.gui_prop_name_ring,
        )

        // Create a new prop of the given type.
        @Throws(JuggleExceptionUser::class)
        fun newProp(type: String): Prop {
            if (type.equals("ball", ignoreCase = true)) {
                return BallProp()
            } else if (type.equals("image", ignoreCase = true)) {
                return ImageProp()
            } else if (type.equals("ring", ignoreCase = true)) {
                return RingProp()
            }

            val message = getStringResource(Res.string.error_prop_type, type)
            throw JuggleExceptionUser(message)
        }

        val COLOR_NAMES: List<String> = listOf(
            "transparent",
            "black",
            "blue",
            "cyan",
            "gray",
            "green",
            "magenta",
            "orange",
            "pink",
            "red",
            "white",
            "yellow",
        )

        val COLOR_VALS: List<Color> = listOf(
            Color(0x00000000),
            Color.Black,
            Color.Blue,
            Color.Cyan,
            Color.Gray,
            Color.Green,
            Color.Magenta,
            Color(0xFFFFC800),
            Color(0xFFFFAFAF),
            Color.Red,
            Color.White,
            Color.Yellow,
        )

        val COLOR_MIXED: List<String> = listOf(
            "red",
            "green",
            "blue",
            "yellow",
            "cyan",
            "magenta",
            "orange",
            "pink",
            "gray",
            "black",
        )
    }
}
